### PR TITLE
Ensure WebDAV clobbering protection retrieves ETag if necessary

### PR DESCRIPTION
**Pull request by FND**
> it appears that some servers do not provide an ETag in the `PUT`
> response, so we have to perform an additional `HEAD` request
>
> cf. 6505e6f44831358d8c163fe5ebe3479537b7209e [#2614]

— bca1a5f5d695efc59a7d2242efd2179a4efd774b

cf. https://github.com/Jermolene/TiddlyWiki5/pull/2614#discussion_r159143156

caveats:

* this particular implementation is currently untested
* I'm not sure whether `HEAD` requests are universally supported, but `GET` seems a bit excessive?
* callback hell makes me shudder, but I'm not sure how to avoid that here

### CODE REVIEW

**Comment by Jermolene** [comment]
Thanks @FND 

You mention that the code is untested, I wonder if @dietmarw or @edrex can help?

**Comment by edrex** [comment]
Thanks @FND for continuing to nudge this forward! I haven't had time to play with TiddyWiki5 in awhile, and testing this could be a fun excuse to poke around. I'll take a look and get back.

**Comment by dietmarw** [comment]
I don't think I can help. Don't even know what this is about tbh ;-)

File: core/modules/savers/put.js [hunk 1]
```diff
@@ -58,21 +58,41 @@ PutSaver.prototype.save = function(text, method, callback) {
...
 	$tw.utils.httpRequest({
-		url: this.uri(),
+		url: uri,
 		type: "PUT",
 		headers: headers,
 		data: text,
 		callback: function(err, data, xhr) {
 			if(err) {
 				callback(err);
 			} if(xhr.status === 200 || xhr.status === 201) {
```
Comment (pmario) [inline core/modules/savers/put.js]: I did just test it with IIS webdav setting described here: https://www.youtube.com/playlist?list=PLuiC_HFhI4OwoVDb-B-VK0ydj-mBPNn-1 

There serve saves the file and returns 204 .... Which is different to the check above, but will invalidate the PR

**Comment by pmario** [comment]
There are really strange things going on. 

 - TW loads with a GET requste that returns an ETAG eg: 1234
 - Requesting HEAD needs {"Accept":"*/*"} otherwise I get 406.
   - HEAD is requested with IF-none-match: 1234
   - HEAD gets an ETAG that is different to GET eg: 4321
 - PUT uses 4321 in If-Match header, which causes 412 :/
 - If HEAD is requested without If-none-match header it retursn ETAG: 1234
    - ETAG 1234 is the one from GET 

I'm confused :/

**Comment by edrex** [comment]
It feels like this needs tests with a mocked server providing the various response cases. Is there precedent for that in the TW5 test suite? Looking..

**Comment by pmario** [comment]
Digging into the topic, for me it seems the ETag is more of a problem, than a solution.
see: https://www.blueclaw.co.uk/2017/05/19/what-are-etags/

I found an interesting resource: https://adaptpartners.com/technical-seo/how-to-implement-page-speed-recommendations-at-server-level/  and have a closer look.

**Comment by icarito** [comment]
I tried a fix that works for me on #3230 - but it seems this is stuck trying to support @pmario 's IIS server. @pmario could you provide a test server setup so that we could try this and make it work for everyone?

My reference DAV server very simple to setup (one command line) is WsgiDAV:

```
pip install wsgidav
wsgidav -r . -p 8000 --server wsgiref
```

Will serve the current directory over WebDAV. I guess tests could be written against it or a node alternative.

**Comment by pmario** [comment]
>@pmario could you provide a test server setup so that we could try this and make it work for everyone?

I don't have the possibility to make an IIS server globally available. I only use it in a local setting. (As a possible replacement to TiddlyFox AddOn)

There was a discussion going on, where I did remember, that every windows user has a IIS license "[out of the box](https://groups.google.com/d/msg/tiddlywiki/8HB0R_ADYac/lqrPlsZ0BwAJ )" 

I did some WebDav experiments and some [intro videos](https://groups.google.com/d/msg/tiddlywiki/_YwmiKqMyrI/6SkubfoQCgAJ), so others are able to use the mechanism. ..

.. In the initial setup / videos I didn't activate Etags. ... But this created a problem, if you write from 2 open windows. 

I'm not 100% sure, how Etags are created with IIS and other systems. 
I'm also not sure, if we should request HEAD, if GET already returns an Etag. 

I'm also not sure, if an index.html created with node is byte by byte compatible with index.html created from the browser save. 

1 workflow is like this: 

 - create index.html from CLI with nodejs
 - --output directory is IIS webroot
 - load TW into browser from `localhost`
 - no changes
 - save -> ETag ERROR

**Comment by pmario** [comment]
@FND could you try this code with your setup?  https://github.com/pmario/TiddlyWiki5/blob/fix-3231/core/modules/savers/put.js ??

**Comment by icarito** [comment]
On 02/05/18 05:00, Mario Pietsch wrote:

> There was a discussion going on, where I did remember, that every
> windows user has a IIS license "out of the box
> <https://groups.google.com/d/msg/tiddlywiki/8HB0R_ADYac/lqrPlsZ0BwAJ>"
>

I have no access to a Windows computer.

> .. In the initial setup / videos I didn't activate Etags. ... But this
> created a problem, if you write from 2 open windows.

Yes I appreciate the ETag support - it allows me to not accidentally
overwrite when working in more than one window.

But it is necessary for it to be robust. Currently it fails for Apache
and WSGIDav users, who don't get ETags from GET and PUT requests.

I don't know which is up to spec but we should avoid to rely on
Microsoft extensions to the standard.

**Comment by pmario** [comment]
>I don't know which is up to spec but we should avoid to rely on Microsoft extensions to the standard.

That's not what I suggested. The HEAD request is fine!

Since the first GET is done from the browser, I don't know a way, how to access this Etag within the webdav save code. It may be available to the browser only. So imo we have to do the HEAD request. 

Please check my changes to your code, if it works with your setup. .. It works for me with win10 IIS10

**Comment by pmario** [comment]
@FND could you try this code with your setup? https://github.com/pmario/TiddlyWiki5/blob/fix-3231/core/modules/savers/put.js

It seems to work for icarito

**Comment by FND** [comment]
I'm afraid I can't, as I don't have access to a suitable WebDAV instance right now. If you could provide me with an empty TiddlyWiki containing this version, I could probably try it next week or so.

It's not clear to me (from the discussion or code) what's actually changed in comparison to the change proposed in this PR?

**Comment by pmario** [comment]
@FND  here is an empy.html that contains the discussed code. [index.zip](https://github.com/Jermolene/TiddlyWiki5/files/1971756/index.zip)

I did test it with win10 IIS, It seems to work for "icarito" and I did a test with an ubuntu nginx webdav installation. It works there. ... At the moment I try to get an alpine apache setup going. .. If it works with those different combinations, we should be save :)

**Comment by FND** [comment]
I'm afraid that doesn't work for me, @pmario: The server (Apache/2.4.25 on Debian) responds with 412 on the very first save request.

```
200 GET tiddlywiki.html
200 OPTIONS tiddlywiki.html
200 HEAD tiddlywiki.html
412 PUT tiddlywiki.html
```

**Comment by pmario** [comment]
How does your URL look like. .. Do you have eg. `http:/localhost/tiddlywiki.html`?

**Comment by pmario** [comment]
412 is strange I always got 409

**Comment by pmario** [comment]
from https://tools.ietf.org/html/rfc7232#section-4.2 ... 

```
   The 412 (Precondition Failed) status code indicates that one or more
   conditions given in the request header fields evaluated to false when
   tested on the server.  This response code allows the client to place
   preconditions on the current resource state (its current
   representations and metadata) and, thus, prevent the request method
   from being applied if the target resource is in an unexpected state.
```

Can you have a look what's different between your Request Header that works and the new one?

**Comment by FND** [comment]
412 is what happens if the `If-Match` header contains a different ETag. However, it appears TiddlyWiki is doing the right thing there (`PUT`ing with the ETag previously retrieved via `HEAD`), so I'm unsure why Apache rejects that. 🤷‍♂️

> How does your URL look like.

I assume you're asking because of your [issue with implicit file names](https://github.com/Jermolene/TiddlyWiki5/pull/3230#issuecomment-386533750): That's not the case for me, I used https://…/tiddlywiki.html here (I'm not surprised implicit file names don't work as you expected).

**Comment by FND** [comment]
Found the issue: Maddeningly, it appears Apache adds `-gzip` to the end of ETag if GZIP compression is enabled, but does not account for that suffix when validating `If-Match` headers.

So anyway, it appears your sample file works after all. Since #3230 has been merged, I guess we can close this now. (I'd still be curious what's different in that implementation, since it's not obvious.)

**Comment by pmario** [comment]
So it seems creating ETags is a "dark art" :)

**Comment by pmario** [comment]
And consulting the web .. "non-local" WebDAV seems to be a security problem :/

**Comment by FND** [comment]
Thanks for your explanations, @pmario, that resolves some of my confusion.

> IIS `HEAD` request needs the Accept-header to be `*/*` otherwise I did get `406 Not Acceptable`
> The old Accept header seems to be `application/json` ... may be left over from `OPTIONS` request

I was wondering about that: Since `HEAD` requests should never contain a response body, the content type should not make any difference and thus `Accept: */*` is confusing.

TiddlyWiki's `httpRequest` [defaults to JSON](https://github.com/Jermolene/TiddlyWiki5/blob/f9be417204695d26d5f458fbc53870b2adb9a372/core/modules/utils/dom/http.js#L24), so `headers: {}` should suffice there.

> the HEAD request failed (406) the consecutive PUTs where executed without an ETag. ... So they succeeded.

That sounds like a potentially dangerous bug (because it might lead to undesirable clobbering): If the `HEAD` request fails, the `PUT` should not be initiated in the first place.

>  After a successful PUT the following HEAD request returns a "weak" ETag like -> W/"123123123"
> Which needs to be filtered out because PUT seems to need a "strong" ETag

That was another thing that confused me. I'm somewhat appalled that this is necessary (because ETags should be considered opaque, i.e. we shouldn't modify them on our end), but if that's what it takes...

> consulting the web .. "non-local" WebDAV seems to be a security problem :/

I'm not sure why anyone would assert that. I'm not aware of any WebDAV-specific security concerns, other than the obvious fact that if access is not restricted (e.g. via a password), anyone could write to your drive.

**Comment by pmario** [comment]
>That was another thing that confused me. I'm somewhat appalled that this is necessary (because ETags should be considered opaque, i.e. we shouldn't modify them on our end), but if that's what it takes...

You are right, that's probably "fighting the symptoms" .... But I don't have much experience with ETags, and I couldn't find any info how IIS actually creates it.

**Comment by pmario** [comment]
>That sounds like a potentially dangerous bug (because it might lead to undesirable clobbering): If the HEAD request fails, the PUT should not be initiated in the first place.

But then we need much better "user feedback" ... As a developer I hit F12 and have a look, what's going on. .. But as a User I just stop using "the crap" if it doesn't do, what I expect. .. 

So we would need some more "upfront" info. Eg: 

 - Your WebDav server, doesn't create an ETag. Saving with this window, may overwrite changes someone else made! Contact your IT :p
 - The HEAD request failed: .... some reasoning and info what user should do?
    - Instead of HEAD we should use a term, that a user understands
 - 409-conflict ... Make sure your URL contains `https://.../index.html` or what ever name should be there. 

and so on. ..

**Pull request closed without merging**